### PR TITLE
Scheduler config validation

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
As of now, when feature gates are enabled, we are just removing the unnecessary predicate keys from predicateMap like this https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/algorithmprovider/defaults/defaults.go#L185, the problem is we are calling 
https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/algorithmprovider/plugins.go from  https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-scheduler/app/server.go#L131and after that, say if we load config from policy configmap or policy config file for scheduler we have no knowledge of why a predicate is removed from predicateMap with scheduler starting and failing with errors like 

`I0829 07:09:12.059949       1 factory.go:977] Registering predicate: CheckNodeMemoryPressure
F0829 07:09:12.059991       1 plugins.go:229] Invalid configuration: Predicate type not found for CheckNodeMemoryPressure` 

which has no context on why CheckNodeMemoryPressure has been removed from predicate. In this case we are seeing this error because we have enabled TaintNodesByCondition using featuregates which removes CheckNodeMemoryPressure and we have asked to load CheckNodeMemoryPressure from policy config file

- This PR adds validation logic to ensure that there is a mapping between featuregates and corresponding predicates and priorities needed. After this PR, when we have a misconfiguration from policy config file, we would see an error like:

`[Predicate CheckNodeMemoryPressure shouldn't be in when TaintNodesByCondition is enabled, Predicate PodToleratesNodeTaints should be in when TaintNodesByCondition is enabled]`

This ensures dependency between featuregates and their needed, not needed predicates/priorities.

xref: https://bugzilla.redhat.com/show_bug.cgi?id=1623346

/sig scheduling
/cc @bsalamat @k82cn @Huang-Wei @aveshagarwal

### CODE REVIEW

File: pkg/scheduler/api/types.go [hunk 1]
```diff
@@ -309,6 +309,15 @@ type HostPriority struct {
...
 }
 
+// FeatureDependencies provides the feature and its dependent predicates and priorities.
+type FeatureDependency struct {
+	Name                  string   // Name of the feature gate.
+	NeededPredicateList   []string // List of predicates needed if we enable this predicate or priority
+	NeededPriorityList    []string // List of priorities needed if we enable this predicate or priority.
+	ExcludedPredicateList []string // List of predicates that should be excluded if we enable this predicate/priority.
+	ExcludedPriorityList  []string // List of priorities that should be excluded if we enable this predicate/priority
+}
```
Comment (Huang-Wei) [inline pkg/scheduler/api/types.go]: Maybe change `[]string` to `map[string]struct{}` so searching will be more performant - from O(n) to O(1). And also avoid duplication.

File: pkg/scheduler/api/validation/validation.go [hunk 2]
```diff
@@ -28,16 +28,68 @@ import (
...
 
+// checkIfExists checks if the given name exists in the string slice.
+func checkIfExistsInList(name string, nameList []string) bool {
+	for _, itemName := range nameList {
+		if name == itemName {
+			return true
+		}
+	}
+	return false
+}
```
Comment (Huang-Wei) [inline pkg/scheduler/api/validation/validation.go]: It's possible to make it a O(1) op (see https://github.com/kubernetes/kubernetes/pull/68103/commits/22242a072ef94ddac69771ff91fe9c5864f617d1#r214219724)

File: pkg/scheduler/algorithmprovider/defaults/defaults.go [hunk 3]
```diff
@@ -195,21 +198,28 @@ func ApplyFeatureGates() {
...
 
+		excludedPredicates = append(excludedPredicates, predicates.CheckNodeConditionPred, predicates.CheckNodeMemoryPressurePred, predicates.CheckNodeDiskPressurePred, predicates.CheckNodePIDPressurePred)
 		// Fit is determined based on whether a pod can tolerate all of the node's taints
 		factory.RegisterMandatoryFitPredicate(predicates.PodToleratesNodeTaintsPred, predicates.PodToleratesNodeTaints)
 		// Insert Key "PodToleratesNodeTaints" and "CheckNodeUnschedulable" To All Algorithm Provider
 		// The key will insert to all providers which in algorithmProviderMap[]
 		// if you just want insert to specific provider, call func InsertPredicateKeyToAlgoProvider()
 		factory.InsertPredicateKeyToAlgorithmProviderMap(predicates.PodToleratesNodeTaintsPred)
 
+		// TODO: @ravig - Make this as set instead of list so that we can avoid duplication.
```
Comment (ravisantoshgudimetla) [inline pkg/scheduler/algorithmprovider/defaults/defaults.go]: The set here is a created from a map, so O(1) is very much possible.

File: pkg/scheduler/api/types.go (hunk 1 above)
Comment (ravisantoshgudimetla) [reply pkg/scheduler/api/types.go]: Yes, ofcourse, that is what I meant here - https://github.com/kubernetes/kubernetes/pull/68103/files#r214220679 and in the TODO:

File: pkg/scheduler/algorithmprovider/defaults/defaults.go [hunk 4]
```diff
@@ -178,10 +179,12 @@ func defaultPredicates() sets.String {
 }
 
 // ApplyFeatureGates applies algorithm by feature gates.
-func ApplyFeatureGates() {
+func ApplyFeatureGates() []api.FeatureDependency {
+	featureDependency := []api.FeatureDependency{}
```
Comment (Huang-Wei) [inline pkg/scheduler/algorithmprovider/defaults/defaults.go]: If it's required to search by featureGate name, I prefer a map (or DAG) over slice.

But I looked through the code, it seems it's just iterating the `[]FeatureDependency` using `range()` in ValidatePolicy(), so this looks fine to me - don't do early-optimization :grin:

File: pkg/scheduler/api/validation/validation.go [hunk 5]
```diff
@@ -28,16 +28,68 @@ import (
...
+			}
+		}
+	}
+	for _, schedulerDependency := range featureDependencies {
+		if len(schedulerDependency.NeededPredicateList) > 0 {
+			for _, neededPredicate := range schedulerDependency.NeededPredicateList {
+				found := false
+				for _, predicate := range policy.Predicates {
+					if predicate.Name == neededPredicate {
+						found = true
```
Comment (Huang-Wei) [inline pkg/scheduler/api/validation/validation.go]: Based on current logic, should add a `break` underneath to jump out early.

File: pkg/scheduler/api/validation/validation.go [hunk 6]
```diff
@@ -28,16 +28,68 @@ import (
...
+				}
+			}
+
+		}
+		if len(schedulerDependency.NeededPriorityList) > 0 {
+			for _, neededPriority := range schedulerDependency.NeededPriorityList {
+				found := false
+				for _, priority := range policy.Priorities {
+					if priority.Name == neededPriority {
+						found = true
```
Comment (Huang-Wei) [inline pkg/scheduler/api/validation/validation.go]: Based on current logic, should add a `break` underneath to jump out early.

File: pkg/scheduler/api/validation/validation.go [hunk 7]
```diff
@@ -28,16 +28,68 @@ import (
...
+					}
+				}
+				if !found {
+					validationErrors = append(validationErrors, fmt.Errorf("Priority %s should be in when %s is enabled", neededPriority, schedulerDependency.Name))
+				}
+			}
+
+		}
+	}
```
Comment (Huang-Wei) [inline pkg/scheduler/api/validation/validation.go]: Line 51 to 92 seems to validate:

1. each NeededXYZ is in Policy.XYZ
1. each Policy.XYZ is not in ExcludedXYZ

XYZ = Predicates/Priorities

Some suggestions:
1. overall time complexity can be reduced if we change NeededXYZ from `[]string` to `map[string]struct{}`. but seems policy.XYZ is also a slice, maybe need to refactor a little bit also to ensure the operation "each Policy.XYZ is not in ExcludedXYZ" is also O(1)
1. `if len(...) > 0` is unnecessary

File: pkg/scheduler/api/validation/validation.go (hunk 5 above)
Comment (ravisantoshgudimetla) [reply pkg/scheduler/api/validation/validation.go]: I think, we should aggregate over the entire list of failed predicates and priorities and then bubble it up to the user.

**Comment by bsalamat** [comment]
/remove-lifecycle rotten
/remove-lifecycle stale

@ravisantoshgudimetla Do you want to update and keep this PR?

**Comment by ravisantoshgudimetla** [comment]
@bsalamat - Yes I will look into this for 1.15.

**Pull request closed without merging**